habitat-hitl/habitat_hitl/_internal/networking/networking_process.py

```diff
@@ -140,155 +200,252 @@ async def receive_client_states(self, websocket: ClientConnection) -> None:
+                        except Exception as e:
+                            print(f"Error sending to client. Error: {e}.")
+                            self.handle_disconnect(slot.connection_id)
+
+            time_end_ns = time.time_ns()
+            elapsed_s = (time_end_ns - time_start_ns) / (10**9)
+            keyframe_send_sleep_time = 0.02
+            wait_time = keyframe_send_sleep_time - min(
+                elapsed_s, keyframe_send_sleep_time
+            )
+            await asyncio.sleep(wait_time)
 
     def has_connection(self) -> bool:
         return len(self._connected_clients) > 0
 
-    def handle_disconnect(self) -> None:
+    def can_accept_connection(self) -> bool:
+        return (
+            len(self._connected_clients)
+            < self._networking_config.max_client_count
+        )
+
+    def handle_disconnect(self, connection_id: int) -> None:
         """
         To be called after a websocket has closed. Don't call this to close the websocket.
         """
         if len(self._connected_clients) == 0:
             return
-        assert len(self._connected_clients) == 1
-        websocket_id = list(self._connected_clients.keys())[0]
-        websocket = self._connected_clients[websocket_id]
-        # todo: assert that websocket is actually already closed
+
+        # TODO: Connections can sometimes be closed multiple times.
+        if connection_id not in self._connected_clients:
+            print("Already disconnected.")
+            return
+
+        # Ensure that the connection is closed.
+        websocket = self._connected_clients[connection_id]
+        asyncio.create_task(websocket.close())
+
         print(f"Closed connection to client  {websocket.remote_address}")
-        del self._connected_clients[websocket_id]
+        del self._connected_clients[connection_id]
+
+        # TODO: Consolidate these states.
+        found = False
+        for user_index, client in self._user_slots.items():
+            if client.connection_id == connection_id:
+                self._free_user_slot(user_index)
+                found = True
+                break
+
+        assert found
 
         disconnection_record: DisconnectionRecord = {}
-        disconnection_record["connectionId"] = websocket_id
+        disconnection_record["connectionId"] = connection_id
+        disconnection_record["userIndex"] = user_index
+        disconnection_record["timestamp"] = str(int(time.time()))
         self._interprocess_record.send_disconnection_record_to_main_thread(
             disconnection_record
         )
 
     def parse_connection_record(self, message: str) -> ConnectionRecord:
         connection_record: ConnectionRecord = json.loads(message)
-        if "isClientReady" not in connection_record:
-            raise ValueError(
-                "isClientReady key not found in initial client message."
-            )
+        # Hack: connection record may be missing. Try to get it from client state.
```

Comment:
This reliably occurs on AWS, but not on local network.

I haven't had the time to go to the bottom of this.